File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/SavingThrowEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum SavingThrowEnum {
    FORTITUDE("fortitude"),
    REFLEX("reflex"),
    WILL("will");

    private final String value;
    private final static Map<String, SavingThrowEnum> CONSTANTS = new HashMap<>();

    static {
        for (SavingThrowEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    SavingThrowEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static SavingThrowEnum fromValue(String value) {
        SavingThrowEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/SavingThrowEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class SavingThrowEnum(private val value: String) {
    FORTITUDE("fortitude"),
    REFLEX("reflex"),
    WILL("will");

    companion object {
        private val CONSTANTS: MutableMap<String, SavingThrowEnum> = HashMap()

        @JsonCreator
        fun fromValue(value: String?): SavingThrowEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[c.value] = c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}